pkg/reconciler/instances/istio/action.go

```diff
@@ -23,20 +40,174 @@ func (a *ReconcileAction) Run(version, profile string, config map[string]interfa
 		return err
 	}
 
-	performer, err := actions.NewDefaultIstioPerformer(context.KubeClient.Kubeconfig(), manifest.Manifest, context.KubeClient, context.Logger, &istioctl.DefaultCommander{})
+	ver, err := a.performer.Version(context.WorkspaceFactory, version, istioChart, context.KubeClient.Kubeconfig(), context.Logger)
+	if err != nil {
+		return errors.Wrap(err, "Could not fetch Istio version")
+	}
+
+	context.Logger.Infof("Detected versions: istioctl %s, pilot version: %s, data plane version: %s", ver.ClientVersion, ver.PilotVersion, ver.DataPlaneVersion)
+
+	if shouldInstall(ver) {
+		context.Logger.Info("No Istio version was detected on the cluster, performing installation...")
+
+		err = a.performer.Install(context.KubeClient.Kubeconfig(), manifest.Manifest, context.Logger)
+		if err != nil {
+			return errors.Wrap(err, "Could not install Istio")
+		}
+
+		err = a.performer.PatchMutatingWebhook(context.KubeClient, context.Logger)
+		if err != nil {
+			return errors.Wrap(err, "Could not patch MutatingWebhookConfiguration")
+		}
+	} else {
+		if !isClientVersionAcceptable(ver) {
+			context.Logger.Errorf("Istio could not be updated since the binary version: %s is not up to date: %s", ver.ClientVersion, ver.TargetVersion)
+			return nil
+		}
+
+		if isDowngrade(ver) {
+			context.Logger.Errorf("Downgrade detected from version: %s to version: %s - finishing...", ver.PilotVersion, ver.TargetVersion)
+			return nil
+		}
+
+		if !canUpdate(ver) {
+			context.Logger.Errorf("Istio could not be updated from version: %s to version: %s - upgrade can be formed by the most one minor version", ver.PilotVersion, ver.TargetVersion)
+			return nil
+		}
+
+		if isMismatchPresent(ver) {
+			context.Logger.Warnf("Istio components version mismatch detected: pilot version: %s, data plane version: %s", ver.PilotVersion, ver.DataPlaneVersion)
+		}
+
+		context.Logger.Infof("Istio version was detected on the cluster, updating from %s to %s...", ver.PilotVersion, ver.ClientVersion)
+
+		err = a.performer.Update(context.KubeClient.Kubeconfig(), manifest.Manifest, context.Logger)
+		if err != nil {
+			return errors.Wrap(err, "Could not update Istio")
+		}
+
+		kubeClient, err := retrieveClientsetFrom(context.KubeClient.Kubeconfig(), context.Logger)
+		if err != nil {
+			return err
+		}
+
+		cfg := istioConfig.IstioProxyConfig{
+			ImagePrefix:           istioImagePrefix,
+			ImageVersion:          fmt.Sprintf("%s-distroless", ver.TargetVersion),
+			RetriesCount:          retriesCount,
+			DelayBetweenRetries:   delayBetweenRetries,
+			SleepAfterPodDeletion: sleepAfterPodDeletion,
+			Kubeclient:            kubeClient,
+			Debug:                 true,
+			Log:                   context.Logger,
+		}
+
+		err = a.istioProxyReset.Run(cfg)
```

Comment:
It would be nice to make it more readable, too many "if's". How about encapsulating this in a single method? It's reasonable to assume all theses checks/vlidations are always executed.